deepmd/utils/tabulate.py

```diff
@@ -0,0 +1,451 @@
+                        idx = [
+                            (type_i, ii) not in self.exclude_types
+                            and self.sel_a[type_i] > 0
+                            for type_i in range(self.ntypes)
+                        ]
+                        uu = np.max(upper[idx])
+                        ll = np.min(lower[idx])
+                    else:
+                        ielement = ii // self.ntypes
+                        net = (
+                            "filter_" + str(ielement) + "_net_" + str(ii % self.ntypes)
+                        )
+                        if self.is_pt:
+                            uu = np.max(upper[ielement])
+                            ll = np.min(lower[ielement])
+                        else:
+                            uu = upper[ielement]
+                            ll = lower[ielement]
+                    xx = np.arange(ll, uu, stride0, dtype=self.data_type)
+                    xx = np.append(
+                        xx,
+                        np.arange(uu, extrapolate * uu, stride1, dtype=self.data_type),
+                    )
+                    xx = np.append(
+                        xx, np.array([extrapolate * uu], dtype=self.data_type)
+                    )
+                    nspline = (
+                        (uu - ll) / stride0 + (extrapolate * uu - uu) / stride1
+                    ).astype(int)
+                    self._build_lower(
+                        net, xx, ii, uu, ll, stride0, stride1, extrapolate, nspline
+                    )
+        elif self.descrpt_type == "T":
+            xx_all = []
+            for ii in range(self.ntypes):
+                if self.is_pt:
+                    uu = np.max(upper[ii])
+                    ll = np.min(lower[ii])
+                else:
+                    ll = lower[ii]
+                    uu = upper[ii]
+                xx = np.arange(extrapolate * ll, ll, stride1, dtype=self.data_type)
+                xx = np.append(xx, np.arange(ll, uu, stride0, dtype=self.data_type))
+                xx = np.append(
+                    xx,
+                    np.arange(
+                        uu,
+                        extrapolate * uu,
+                        stride1,
+                        dtype=self.data_type,
+                    ),
+                )
+                xx = np.append(xx, np.array([extrapolate * uu], dtype=self.data_type))
+                xx_all.append(xx)
+            nspline = (
+                (upper - lower) / stride0
+                + 2 * ((extrapolate * upper - upper) / stride1)
+            ).astype(int)
+            idx = 0
+            for ii in range(self.ntypes):
+                for jj in range(ii, self.ntypes):
+                    net = "filter_" + str(ii) + "_net_" + str(jj)
+                    self._build_lower(
+                        net,
+                        xx_all[ii],
+                        idx,
+                        uu,
+                        ll,
```

Comment:
this is wrong! `uu` and `ll` are defined in the previous loop

Reply:
I expect https://github.com/deepmodeling/deepmd-kit/pull/4227/commits/15bf3e2361146496c9201a336096d476aaed776b fixes it.